core/src/main/java/io/grpc/SynchronizationContext.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2018 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A synchronization context is a queue of tasks that run in sequence.  It offers following
+ * guarantees:
+ *
+ * <ul>
+ *    <li>Ordering.  Tasks are run in the same order as they are submitted via {@link #execute}
+ *        and {@link #executeLater}.</li>
+ *    <li>Serialization.  Tasks are run in sequence and establish a happens-before relationship
+ *        between them. </li>
+ *    <li>Non-reentrancy.  If a task running in a synchronization context executes or schedules
+ *        another task in the same synchronization context, the latter task will never run
+ *        inline.  It will instead be queued and run only after the current task has returned.</li>
+ * </ul>
+ *
+ * <p>It doesn't own any thread.  Tasks are run from caller's or caller-provided threads.
+ *
+ * <p>This class is thread-safe.
+ *
+ * @since 1.17.0
+ */
+@ThreadSafe
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4984")
+public final class SynchronizationContext implements Executor {
+  private static final Logger log = Logger.getLogger(SynchronizationContext.class.getName());
+
+  private final Object lock = new Object();
+  private final UncaughtExceptionHandler uncaughtExceptionHandler;
+
+  @GuardedBy("lock")
+  private final Queue<Runnable> queue = new ArrayDeque<Runnable>();
+  @GuardedBy("lock")
+  private boolean draining;
+
+  /**
+   * Creates a SynchronizationContext.
+   *
+   * @param uncaughtExceptionHandler handles exceptions thrown out of the tasks.  Different from
+   *        what's documented on {@link UncaughtExceptionHandler#uncaughtException}, the thread is
+   *        not terminated when the handler is called.
+   */
+  public SynchronizationContext(UncaughtExceptionHandler uncaughtExceptionHandler) {
```

Comment:
We could provide a zero-arg version that just logs by default. But we can do that at any time. This seems fine for now.